controllers/onload_controller.go

```diff
@@ -78,34 +90,515 @@ func (r *OnloadReconciler) Reconcile(
+		if err != nil {
+			log.Error(err, "Failed to patch node"+node.Name)
+			return err
+		}
+
+	}
+	return nil
+}
+
+func (r *OnloadReconciler) addKmmLabelsToNodes(ctx context.Context, onload *onloadv1alpha1.Onload) (*ctrl.Result, error) {
+	log := log.FromContext(ctx)
+	kmmPodLabelSet := labels.Set{"kmm.node.kubernetes.io/module.name": onload.Name + onloadModuleNameSuffix}
+
+	labels := labels.FormatLabels(onload.Spec.Selector)
+	labelKey := kmmLabelName(onload.Name, onload.Namespace)
+
+	nodes, err := r.listNodesWithLabels(ctx, labels)
+	if err != nil {
+		return nil, err
+	}
+
+	changesMade := false
+
+	for _, node := range nodes.Items {
+		if _, found := node.Labels[labelKey]; !found {
+			// Check that there isn't a lingering module pod
+			// This can be removed, but without it the upgrade process is more
+			// concurrent (rather than rolling), which leads to more pod
+			// restarts/failures. Both methods work, but this approach maintains
+			// the rolling nature of the update (as much as is possible).
+			pods, err := r.getPodsOnNode(ctx, kmmPodLabelSet, node.Name)
+			if err != nil {
+				return nil, err
+			} else if len(pods) > 0 {
+				log.Info("lingering module pod on node " + node.Name)
+				return &ctrl.Result{RequeueAfter: defaultRequeueTime}, nil
+			}
+
+			nodeCopy := node.DeepCopy()
+			node.Labels[labelKey] = onload.Spec.Onload.Version
+			err = r.Patch(ctx, &node, client.MergeFrom(nodeCopy))
+			if err != nil {
+				log.Error(err, "Failed to patch node with new label")
+				return nil, err
+			}
+			changesMade = true
+		}
+	}
+
+	var res *ctrl.Result = nil
+	if changesMade {
+		res = &ctrl.Result{Requeue: true}
+	}
+
+	return res, nil
+}
+
+func (r *OnloadReconciler) addOnloadLabelsToNodes(ctx context.Context, onload *onloadv1alpha1.Onload) (*ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
+	labels := labels.FormatLabels(onload.Spec.Selector)
+	labelKey := onloadLabelName(onload.Name, onload.Namespace)
+	kmmLabel := kmmLabelName(onload.Name, onload.Namespace)
+	nodes, err := r.listNodesWithLabels(ctx, labels)
+	if err != nil {
+		return nil, err
+	}
+
+	changesMade := false
+
+	for _, node := range nodes.Items {
+		// Don't add the onload label onto a node either without kmm or with the
+		// wrong onload module version.
+		if node.Labels[kmmLabel] != onload.Spec.Onload.Version {
+			continue
+		}
+		if _, found := node.Labels[labelKey]; !found {
+			nodeCopy := node.DeepCopy()
+			node.Labels[labelKey] = onload.Spec.Onload.Version
+			err := r.Patch(ctx, &node, client.MergeFrom(nodeCopy))
+			if err != nil {
+				log.Error(err, "Failed to patch node with new label",
+					"Node", node.Name)
+				return nil, err
+			}
+			changesMade = true
+		}
+	}
+
+	var res *ctrl.Result = nil
+	if changesMade {
+		res = &ctrl.Result{Requeue: true}
+	}
+
+	return res, nil
+}
+
+func (r *OnloadReconciler) getBadNodes(ctx context.Context, onload *onloadv1alpha1.Onload) ([]corev1.Node, error) {
```

Comment:
Shall we call "bad" nodes something else, like "nodesToUpgrade" (though it's still not the most accurate name)?